Cubical/Homotopy/Connected.agda

```diff
@@ -12,46 +12,123 @@ open import Cubical.Foundations.Path
-  isoToIsEquiv
-    (iso (_∘ f)
-      (λ t b → inv t b (fConn b .fst))
-      (λ t → funExt λ a →
-        cong (inv t (f a)) (fConn (f a) .snd ∣ a , refl ∣)
-        ∙ substRefl {B = fst ∘ P} (t a))
-      (λ s → funExt λ b →
-        Trunc.elim
-          {B = λ d → inv (s ∘ f) b d ≡ s b}
-          (λ _ → isOfHLevelPath n (P b .snd) _ _)
-          (λ {(a , p) i → transp (λ j → P (p (j ∨ i)) .fst) i (s (p i))})
-          (fConn b .fst)))
-  where
-  inv : ((a : A) → P (f a) .fst) → (b : B) → hLevelTrunc n (fiber f b) → P b .fst
-  inv t b =
-    Trunc.rec
-      (P b .snd)
-      (λ {(a , p) → subst (fst ∘ P) p (t a)})
+isHLevelTruncatedFun : ∀ {ℓ ℓ'} (n : ℕ) {A : Type ℓ} {B : Type ℓ'} (f : A → B) → Type (ℓ-max ℓ ℓ')
+isHLevelTruncatedFun n f = ∀ b → isOfHLevel n (fiber f b)
+
+isConnectedSubtr : ∀ {ℓ ℓ'} {A : Type ℓ} {B : Type ℓ'} (n m : ℕ) (f : A → B)
+                → isHLevelConnectedFun (m + n) f
+                → isHLevelConnectedFun n f
+isConnectedSubtr n m f iscon b =
+  transport (λ i → isContr (ua (truncOfTruncEq {A = fiber f b} n m) (~ i) ))
+            (∣ iscon b .fst ∣ ,
+              Trunc.elim (λ x → isOfHLevelPath n (isOfHLevelTrunc n) _ _)
+                     λ a → cong ∣_∣ (iscon b .snd a))
+
+private
+  typeToFiber : ∀ {ℓ} (A : Type ℓ)  (b : Unit) → A ≡ fiber (λ (x : A) → b) b
+  typeToFiber A tt = isoToPath typeToFiberIso
+    where
+    typeToFiberIso : Iso A (fiber (λ (x : A) → tt) tt)
+    Iso.fun typeToFiberIso x = x , refl
+    Iso.inv typeToFiberIso = fst
+    Iso.rightInv typeToFiberIso b i = fst b , (isOfHLevelSuc 1 (isPropUnit) tt tt (snd b) refl) i
+    Iso.leftInv typeToFiberIso a = refl
+
+
+module elim {ℓ ℓ' : Level} {A : Type ℓ} {B : Type ℓ'} (f : A → B) (n : ℕ) where
+
+  isIsoPrecompose : ∀ {ℓ'''} (P : B → HLevel ℓ''' n)
+                   → isHLevelConnectedFun n f
+                   → Iso ((b : B) → P b .fst) ((a : A) → P (f a) .fst)
+  isIsoPrecompose P fConn =
+      (iso (_∘ f)
+        (λ t b → inv t b (fConn b .fst))
+        (λ t → funExt λ a →
+          cong (inv t (f a)) (fConn (f a) .snd ∣ a , refl ∣)
+          ∙ substRefl {B = fst ∘ P} (t a))
+        (λ s → funExt λ b →
+          Trunc.elim
+            {B = λ d → inv (s ∘ f) b d ≡ s b}
+            (λ _ → isOfHLevelPath n (P b .snd) _ _)
+            (λ {(a , p) i → transp (λ j → P (p (j ∨ i)) .fst) i (s (p i))})
+            (fConn b .fst)))
+       where
+    inv : ((a : A) → P (f a) .fst) → (b : B) → hLevelTrunc n (fiber f b) → P b .fst
+    inv t b =
+      Trunc.rec
+        (P b .snd)
+        (λ {(a , p) → subst (fst ∘ P) p (t a)})
+
+  isEquivPrecompose : ∀ {ℓ'''} (P : B → HLevel ℓ''' n)
+                   → isHLevelConnectedFun n f
+                   → isEquiv λ(s : (b : B) → P b .fst) → s ∘ f
+  isEquivPrecompose P fConn = isoToIsEquiv (isIsoPrecompose P fConn)
+
+  isConnectedPrecompose : ((∀ {ℓ'''} (P : B → HLevel ℓ''' n)
+                                  → hasSection λ(s : (b : B) → P b .fst) → s ∘ f))
```

Comment:
Hmm, yes, I remember that I was very annoyed about this too. If I remember correctly, the proof required me to use the assumption for a dependent type of level l''' + 1, which prevented me from fixing the universe level first. It's been quite some time since I wrote this up, so I'll have to double check.